billing.rb

```diff
@@ -0,0 +1,135 @@
+#!/usr/bin/env ruby
+# frozen_string_literal: true
+
+require 'net/http'
+require 'json'
+require 'prometheus/client'
+require 'byebug'
+
+ORG = 'dfe'
+BILLING_URL = 'https://billing.london.cloud.service.gov.uk'
+ORG_GUID = '386a9502-d9b6-4aba-b3c3-ebe4fa3f963e'
+API_URL = 'https://api.london.cloud.service.gov.uk'
+PRECISION = 2
+
+# Rack middleware to fetch billing data from the PaaS billing API and aggregate the data into prometheus metrics
+class BillingCalculator
+  attr_reader :cost
+
+  def initialize(app)
+    @skip_login = false
+    if ENV['SKIP_LOGIN'] && ENV['SKIP_LOGIN'].downcase == 'true'
+      @skip_login = true
+    else
+      @paas_username = ENV.fetch('PAAS_USERNAME')
+      @paas_password = ENV.fetch('PAAS_PASSWORD')
+    end
+
+    @app = app
+    prometheus = Prometheus::Client.registry
+    @cost = prometheus.gauge(
+      :cost,
+      docstring: 'A counter representing PaaS cost per space and resource type on the previous day',
+      labels: %i[space resource_type date]
+    )
+  end
+
+  def update_cost_metric(token)
+    yesterday_billing_data = yesterday_billing_data(token, range_start_yesterday, range_stop_today)
+    global_cost = calculate_cost yesterday_billing_data
+    cost_array = metrics(global_cost).sort_by { |c| "#{c[:space]}-#{c[:resource_type]}" }
+
+    cost_array.each do |c|
+      @cost.set(c[:price], labels: { space: c[:space], resource_type: c[:resource_type], date: range_start_yesterday })
```

Comment:
There's an `attr_reader :cost` so it's better to access through that rather than reference the instance variable directly. You could also add `:app, :pass_username, :pass_password, skip_login` and do the same. 
```suggestion
      cost.set(c[:price], labels: { space: c[:space], resource_type: c[:resource_type], date: range_start_yesterday })
```